SIL: Plumb abstraction patterns through type lowering.

jckarter opened the pull request:
Lowering a SIL type should be a pure function of the formal type of a value and the
abstraction pattern it's being lowered against, but we historically did not carry
enough information in abstraction patterns to lower generic parameter types, so we
relied on a generic context signature that would be pushed and popped before lowering
interface types. This patch largely eliminates the necessity for that, by making it
so that `TypeClassifierBase` and its subclasses now take an `AbstractionPattern`
all the way down, and fixing up the visitor logic so that it derives appropriate
abstraction patterns for tuple elements, function arguments, and aggregate fields too.
This makes it so that type lowering is independent of the current generic context.
(Unfortunately, there are still places scattered across the code where we use the
current generic context in order to build abstraction patterns that we then feed
into type lowering, so we can't yet completely eliminate the concept.)

This then enables us to integrate substituted function type construction into type
lowering as well, since we can now lower a generic parameter type against an
abstraction pattern without that generic parameter having to be tied to the same
generic signature (or any generic signature at all, which in the case of a
substituted function type hasn't necessarily even been finalized yet.)

jckarter committed:
Factor a transform visitor out of IsBindableVisitor.

For substituted function type lowering, we want to take every structural position where a
concrete type matches the original generic type in an interface, and replace it with an
independent generic parameter. The first bit, matching up structural positions to generic
parameters, is more or less what `Type::isBindableTo` already does, so we can adapt its
visitor to take a callback that optionally substitutes into the type being walked based
on the generic argument and binding in the given pair of types.

jckarter commented:
@swift-ci Please test

jckarter commented:
@swift-ci Please test source compatibility

rjmccall commented on lib/SIL/SILFunctionType.cpp:
```diff
@@ -1035,18 +1216,18 @@ static CanSILFunctionType getSILFunctionType(
+    // We don't currently use substituted function types for generic function
+    // type lowering, though we should for generic methods on classes and
+    // protocols.
+    && !genericSig);
```
`subst` seems like a pretty generic name for this variable; maybe `substCollector`, or just anything that suggests it's not a substituted type of some sort?

rjmccall commented on lib/SIL/SILFunctionType.cpp:
```diff
@@ -1056,7 +1237,15 @@ static CanSILFunctionType getSILFunctionType(
+    DestructureResults destructurer(expansionContext, TC, conventions,
+                                    results, subst);
+    destructurer.destructure(origResultType, substFormalResultType);
+  }
```
Out of curiosity, why did the order of destructuring need to change here?  Oh, just so that type parameters in parameters would get named first?

jckarter replied:
Yeah, the exact ordering didn't seem to be important, and I was trying to keep the order of introduced generic parameters for substituted types consistent with what I had implemented before.

rjmccall commented on lib/AST/Type.cpp:
```diff
@@ -1544,33 +1544,41 @@ bool TypeBase::isExactSuperclassOf(Type ty) {
-    IsBindableVisitor() {}
+namespace {
+class IsBindableVisitor
+  : public TypeVisitor<IsBindableVisitor, CanType, CanType>
```
Would you mind splitting the extraction change out from the semantics changes?  There's a lot to pick apart here.

jckarter replied:
OK, I'll try to pull out more separate commits. I might see if I'm able to completely eradicate `getCurGenericContext` while I'm at it.

rjmccall reviewed (commented):
This is a really nice improvement; I love the idea of getting away from the statefulness of the generic context.

jckarter committed:
SIL: Sink GenericContextScope into IRGen.

All the context dependencies in SIL type lowering have been eradicated, but IRGen's
type info lowering is still context-dependent and doesn't systemically pass generic
contexts around. Sink GenericContextScope bookkeeping entirely into IRGen for now.

jckarter commented:
@swift-ci Please test

jckarter commented:
@swift-ci Please test source compatibility

jckarter commented:
@swift-ci Please clean test

jckarter committed:
SIL: Plumb abstraction patterns through type lowering.

Lowering a SIL type should be a pure function of the formal type of a value and the
abstraction pattern it's being lowered against, but we historically did not carry
enough information in abstraction patterns to lower generic parameter types, so we
relied on a generic context signature that would be pushed and popped before lowering
interface types. This patch largely eliminates the necessity for that, by making it
so that `TypeClassifierBase` and its subclasses now take an `AbstractionPattern`
all the way down, and fixing up the visitor logic so that it derives appropriate
abstraction patterns for tuple elements, function arguments, and aggregate fields too.
This makes it so that type lowering is independent of the current generic context.
(Unfortunately, there are still places scattered across the code where we use the
current generic context in order to build abstraction patterns that we then feed
into type lowering, so we can't yet completely eliminate the concept.)

This then enables us to integrate substituted function type construction into type
lowering as well, since we can now lower a generic parameter type against an
abstraction pattern without that generic parameter having to be tied to the same
generic signature (or any generic signature at all, which in the case of a
substituted function type hasn't necessarily even been finalized yet.)

jckarter commented:
@swift-ci Please test

jckarter commented:
@rjmccall I pulled out the function type conversion semantics changes, and went ahead and finished the job of expunging GenericContextScope from SIL (relegating it to IRGen for now, because IRGen's type converter still doesn't plumb through generic signature info at all). What's here now should be NFC. Is this easier to follow?

jckarter commented:
@swift-ci Please test source compatibility

rjmccall commented:
Thanks, LGTM.

jckarter commented:
Thanks John!
